### PR TITLE
Granular Geant setup

### PR DESCRIPTION
**Pull request by sethrj**
Move more of the physics setup out of the demo IO routines into the geant setup, so that the importing is done based on what's loaded by geant in the user code to ensure the greatest consistency.

### CODE REVIEW

File: src/celeritas/ext/GeantSetup.cc [hunk 1]
```diff
@@ -34,50 +31,35 @@ namespace
 {
 //---------------------------------------------------------------------------//
 /*!
- * Set up global geant4 physics.
+ * Load the detector geometry from a GDML input file.
  */
-void load_physics(const GeantSetupOptions& options, G4RunManager* run_manager)
+class DetectorConstruction : public G4VUserDetectorConstruction
```
Comment (stognini) [inline src/celeritas/ext/GeantSetup.cc:36]: Any reason to move the `DetectorConstruction` class inside `GeantSetup.cc` instead of keeping it separate---like `ActionInitialization`, `PrimaryGeneratorAction`, and so on?

Comment (sethrj) [reply src/celeritas/ext/GeantSetup.cc:36]: Only because I was modifying it, saw that:
- it's only used in one translation unit
- it's only a handful of lines
- it's not independently tested

The others you mention are likewise very small, so I'll just do the same :)

File: src/celeritas/ext/detail/GeantPhysicsList.cc [hunk 2]
```diff
@@ -178,122 +198,78 @@ void GeantPhysicsList::add_e_processes()
...
-
-            // Positron
-            auto wentzelvi_msc_positron = std::make_unique<G4WentzelVIModel>();
-            wentzelvi_msc_positron->SetLowEnergyLimit(msc_energy_limit);
-            msc_positron->SetEmModel(wentzelvi_msc_positron.release());
-
-            CELER_LOG(debug) << "Loaded multiple scattering with "
-                                "G4WentzelVIModel";
+            auto model = std::make_unique<G4WentzelVIModel>();
+            model->SetHighEnergyLimit(msc_energy_limit);
```
Comment (stognini) [inline src/celeritas/ext/detail/GeantPhysicsList.cc:265]: I think here we should `SetLowEnergyLimit` since it's the high energy msc model,

File: src/celeritas/ext/detail/GeantPhysicsList.cc [hunk 3]
```diff
@@ -178,122 +198,78 @@ void GeantPhysicsList::add_e_processes()
...
+            process->SetEmModel(model.release());
         }
 
-        physics_list->RegisterProcess(msc_electron.release(), electron);
-        physics_list->RegisterProcess(msc_positron.release(), positron);
+        physics_list->RegisterProcess(process.release(), p);
+
+        CELER_LOG(debug) << "Loaded multiple scattering with "
+                            "G4UrbanMscModel";
```
Comment (stognini) [inline src/celeritas/ext/detail/GeantPhysicsList.cc:272]: Given that we have both `MscModelSelection` set up already, this debug msg can be misleading. I suggest adding a specific message for each msc model loaded.

File: src/celeritas/ext/GeantImporter.cc [hunk 4]
```diff
@@ -266,14 +266,38 @@ store_materials(GeantImporter::DataSelection::Flags particle_flags)
     std::vector<ImportMaterial> materials;
     materials.resize(g4production_cuts_table.GetTableSize());
 
-    std::vector<G4ProductionCutsIndex> particle_cuts;
+    using CutRange = std::pair<G4ProductionCutsIndex,
+                               std::unique_ptr<G4VRangeToEnergyConverter> >;
+
+    std::vector<CutRange> cut_converters;
```
Comment (stognini) [inline src/celeritas/ext/GeantImporter.cc:272]: Is there an advantage in creating a `vector<pair<...>>` instead of `map<...>`?

File: src/celeritas/ext/detail/GeantPhysicsList.cc (hunk 2 above)
Comment (sethrj) [reply src/celeritas/ext/detail/GeantPhysicsList.cc:265]: Oh man, that's my bad. Good catch! And surprising/disturbing that our stepper tests didn't fail as a result...

File: src/celeritas/ext/detail/GeantPhysicsList.cc (hunk 3 above)
Comment (sethrj) [reply src/celeritas/ext/detail/GeantPhysicsList.cc:272]: Good point.

File: src/celeritas/ext/GeantImporter.cc (hunk 4 above)
Comment (sethrj) [reply src/celeritas/ext/GeantImporter.cc:272]: Not meaningfully: I was still thinking of this in terms of "iterate over the ordered g4 indices" as a vector with elements removed, but a map would work just as fine (though it takes up slightly more memory and will be slightly slower, it's not going to be meaningfully different for this case).

File: src/celeritas/ext/GeantPhysicsOptions.hh [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+//! \file celeritas/ext/GeantPhysicsOptions.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "celeritas/Quantities.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+//! Brems selection (TODO: make bitset)
```
Comment (stognini) [inline src/celeritas/ext/GeantPhysicsOptions.hh:15]: I noticed the use of bit flags in `GeantImporter.hh` and wanted to ask: What is the real advantage of setting up `0xN` enums?

File: src/celeritas/ext/GeantImporter.cc (hunk 4 above)
Comment (sethrj) [reply src/celeritas/ext/GeantImporter.cc]: Thinking about this: the way it's used (or intended to be used) matters. We only ever iterate over this container, so it could be as simple as a list. If we needed fast indexed access, it needs to be a vector. If we want fast key-value access, it should be a map. Since `std::list` doesn't have any advantages over small vectors, it's better to leave this as a vector.

File: src/celeritas/ext/GeantPhysicsOptions.hh (hunk 5 above)
Comment (sethrj) [reply src/celeritas/ext/GeantPhysicsOptions.hh:15]: `0xN` is a canonical way of defining bit flags in C and C++ since it's easier to see that `0x00400000u` has only one bit set (compared to `4194304`). In our case we're only using a few bits so it in practice doesn't matter, but I find it to be a nice way of "annotating" that these values aren't meant to be consecutive integers but rather exclusive bit patterns. The "best" way of doing this is to have a code abstraction that lets us turn consecutive enums into a nice bitfield, but I didn't think it worth the effort for this use case so far.

File: src/celeritas/ext/detail/GeantPhysicsList.cc [hunk 6]
```diff
@@ -178,122 +198,81 @@ void GeantPhysicsList::add_e_processes()
...
+        auto process = std::make_unique<G4eMultipleScattering>();
 
-            // Positron
-            auto urban_msc_positron = std::make_unique<G4UrbanMscModel>();
-            urban_msc_positron->SetHighEnergyLimit(msc_energy_limit);
-            msc_positron->SetEmModel(urban_msc_positron.release());
+        if (options_.msc == MscModelSelection::urban)
+        {
+            auto model = std::make_unique<G4UrbanMscModel>();
+            model->SetLowEnergyLimit(msc_energy_limit);
```
Comment (stognini) [inline src/celeritas/ext/detail/GeantPhysicsList.cc:258]: I think you mistakenly changed this bit with an automatic find/replace. Urban has a high limit threshold, whereas wentzelvi has a low limit one.

**Comment by stognini** [comment]
I'm still checking what's up, but my `celeritas_global_AlongStep` test just failed on my mac build (and seems to be a problem in jenkins too).

File: src/celeritas/ext/detail/GeantPhysicsList.cc (hunk 6 above)
Comment (sethrj) [reply src/celeritas/ext/detail/GeantPhysicsList.cc:258]: Ah, that was my error: I didn't notice when updating them in the first place that the two set points were different function calls.

Comment (sethrj) [reply src/celeritas/ext/detail/GeantPhysicsList.cc:258]: Wait, so `msc energy limit` is the boundary *between* two separate MSC models?? What if you only have one or the other enabled?

**Comment by stognini** [comment]
> I'm still checking what's up, but my `celeritas_global_AlongStep` test just failed on my mac build (and seems to be a problem in jenkins too).

Okay, the test was failing because of the energy threshold in the msc urban model. Changing `GeantPhysicsList.cc::258` to
```c++
model->SetHighEnergyLimit(msc_energy_limit);
```
fixed it.

File: src/celeritas/ext/detail/GeantPhysicsList.cc (hunk 6 above)
Comment (stognini) [reply src/celeritas/ext/detail/GeantPhysicsList.cc:258]: Yes. That value is the boundary and by default is 100 MeV.

> What if you only have one or the other enabled?

Good point. I guess we could call `model->SetHighEnergyLimit(100 * TeV)` or similar when enabling only urban, specially since wentzelvi is more important for precise muon/hadron msc at low angles (and single scattering is used for large angles, as far as I remember). 

@whokion any thoughts on this? Do we have a hardcoded limit on your msc implementation in Celeritas?

Comment (stognini) [reply src/celeritas/ext/detail/GeantPhysicsList.cc]: I discussed this point with @whokion offline. He did not explicitly applied a 100 MeV threshold on the Celeritas side, but the model should effectively work only up to this limit and we should add it in the near future (or build urban msc_params below this limit). 

TLDR: his conclusion is that keeping `model->SetHighEnergyLimit(msc_energy_limit)` is better, as we won't have urban msc enabled above the default 100 MeV.

Comment (sethrj) [reply src/celeritas/ext/detail/GeantPhysicsList.cc]: > Use new parameter MscEnergyLimit() (default is 100 MeV) in EM physics
    constructors, allowing to customise transition from low-energy multiple
    scattering to high energy combination of multiple and single scattering.
 
 Ughhh why would you name a *boundary* between two MSC models as an MSC *limit*...

Comment (stognini) [reply src/celeritas/ext/detail/GeantPhysicsList.cc]: Well... a boundary is technically a limit... :grimacing:
But yeah, I totally agree that it's a poorly chosen variable/function naming, and I also got confused the first time I looked into it.

Comment (sethrj) [reply src/celeritas/ext/detail/GeantPhysicsList.cc]: it's not a limit for MSC 🙄 if you called it LowEnergyMscUpperLimit or HighEnergyMscUpperLimit you could argue it... ;)

**Review by stognini — APPROVED** [review]

**Comment by sethrj** [comment]
FIXME: removal of msc args should've been part of the merge commit

File: src/celeritas/ext/detail/GeantPhysicsList.cc [hunk 7]
```diff
@@ -158,16 +180,14 @@ void GeantPhysicsList::add_gamma_processes()
...
  */
-void GeantPhysicsList::add_e_processes()
+void GeantPhysicsList::add_e_processes(G4ParticleDefinition* p)
 {
-    auto       physics_list = G4PhysicsListHelper::GetPhysicsListHelper();
-    const auto electron     = G4Electron::Electron();
-    const auto positron     = G4Positron::Positron();
+    auto* physics_list = G4PhysicsListHelper::GetPhysicsListHelper();
 
-    if (true)
```
Comment (stognini) [inline src/celeritas/ext/detail/GeantPhysicsList.cc:167]: I noticed you removed the `if (true)` from this bit, which made me think about all these `if (true)` statements that have been here for awhile as placeholders...
Do we want to keep them up for this PR, or should we expand the `GeantPhysicsOptions` to include the processes already?

Comment (sethrj) [reply src/celeritas/ext/detail/GeantPhysicsList.cc:167]: Ah good point: I could make it `if (p == G4Positron::Positron() && true)`? I'd like to basically merge the "geant4 benchmark" physics setup routines into our GeantSetup, so all of the physics here could become on/off.

Comment (stognini) [reply src/celeritas/ext/detail/GeantPhysicsList.cc:167]: Exactly. I thought the same, and since we can either use the exported root file from benchmarks or use a manually selected set of options for the `celer-export-geant`, at the end we should have equivalent input options. In the first case, we just check the loaded processes/models, but for the second we will end up with a list of processes to turn on/off. So for now the `if (p == G4Positron::Positron() && true)` is fine, but in the end I think we will end up with something like:

```c++
if (p == G4Positron::Positron() && options_.annihilation)
```

Same goes for the Klein Nishina Compton process, where the `if (true)` was removed.
